### DISCUSSION TITLE
HTTPRoutes (plural) support

### DISCUSSION THREAD

**Discussion by revero-doug** [linkerd/linkerd2#10297] [General]
### What problem are you trying to solve?

`HTTPRoute` precedence rules are arcane at best and difficult to manage declaratively at worst.

to clarify, what I'm proposing would be a higher level CRD/operator that itself manages the construction and application of HTTPRoute manifests, not making an alternative to HTTPRoute resource

### How should the problem be solved?

a `HTTPRoutes` (working title) CRD/operator that allows users to define a series of HTTPRoute specs (or some derivative spec) to explicitly, simply, and declaratively describe route matching precedence would better fit with the overall k8s model of defining systems declaratively and allowing controllers/operators to manage lifecycles. At present, the order in which these resources are applied can affect route matching precedence, which is contrary to these expectations for operating a system on k8s.

The admission controller would only allow one such manifest to target a given `Server`

### Any alternatives you've considered?

maintaining status quo is the only known alternative, and it is problematic for the reason described at the top.

### How would users interact with this feature?

apply the new CRD (working title `HTTPRoutes`) manifest as they would apply any other manifest.

### Would you like to work on this feature?

no

**Comment by kleimkuhler** [comment]
Thanks for the thoughts on this @revero-doug. I understand that matching can be difficult to manage, but Linkerd's HTTPRoute is an attempt to mirror the [Gateway API's HTTPRoute](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.HTTPRouteRule) matching rules; we hope to switch to that CRD in the future.

There has been more discussion about this topic in that repository and we're likely to follow the design decisions that happen upstream, rather than tackling it as part of Linkerd's CRD.

**Comment by revero-doug** [comment]
to clarify, what I'm proposing would be a higher level CRD/operator that itself manages the construction and application of HTTPRoute manifests, not making an alternative to HTTPRoute resource

**Reply by kleimkuhler** [reply]
Okay thanks for clarifying. It seems like something like that could be helpful, but does feel out of scope for Linkerd's development efforts. If you start investigating this yourself though, please keep us updated on how that goes I'd definitely be interested in taking a look!
